packages/core/src/root.ts

```diff
@@ -35,369 +68,711 @@ export function createRemoteRoot<
-  type Text = RemoteText<Root>;
-  type HasChildren = Root | Component;
-  type CanBeChild = Component | Text;
-
-  const parents = new WeakMap<CanBeChild, HasChildren>();
-  const children = new WeakMap<HasChildren, readonly CanBeChild[]>();
-  const props = new WeakMap<Component, any>();
-  const texts = new WeakMap<Text, string>();
-  const tops = new WeakMap<CanBeChild, HasChildren>();
-  const nodes = new WeakSet<CanBeChild>();
 
   let currentId = 0;
-  let mounted = false;
+
+  const rootInternals: RootInternals = {
+    strict,
+    mounted: false,
+    channel,
+    children: EMPTY_ARRAY,
+    nodes: new WeakSet(),
+    parents: new WeakMap(),
+    tops: new WeakMap(),
+  };
 
   const remoteRoot: Root = {
     kind: KIND_ROOT,
     get children() {
-      return children.get(remoteRoot) as any;
+      return rootInternals.children as any;
     },
     createComponent(type, ...rest) {
       if (components && components.indexOf(type) < 0) {
         throw new Error(`Unsupported component: ${type}`);
       }
 
-      let initialProps = rest[0];
-      const initialChildren = rest[1];
+      const [initialProps, initialChildren] = rest;
+
+      const normalizedInitialChildren: AnyChild[] = [];
+      const normalizedInitialProps: {[key: string]: any} = {};
 
       if (initialProps) {
-        // "children" as a prop can be extremely confusing with the "children" of
-        // a component. In React, a "child" can be anything, but once it reaches
-        // a host environment (like this remote `Root`), we want "children" to have
-        // only one meaning: the actual, resolved children components and text.
-        //
-        // To enforce this, we delete any prop named "children". We don’t take a copy
-        // of the props for performance, so a user calling this function must do so
-        // with an object that can handle being mutated.
-        //
-        // I didn’t think checking that the prop exists before deleting it would matter,
-        // but I ran a few benchmarks and it ran substantially faster this way /shrug
-        if (initialProps.children) delete initialProps.children;
-      } else {
-        initialProps = {} as any;
+        for (const key of Object.keys(initialProps)) {
+          // "children" as a prop can be extremely confusing with the "children" of
+          // a component. In React, a "child" can be anything, but once it reaches
+          // a host environment (like this remote `Root`), we want "children" to have
+          // only one meaning: the actual, resolved children components and text.
+          //
+          // To enforce this, we delete any prop named "children". We don’t take a copy
+          // of the props for performance, so a user calling this function must do so
+          // with an object that can handle being mutated.
+          if (key === 'children') continue;
+
+          normalizedInitialProps[key] = makeValueHotSwappable(
+            initialProps[key],
+          );
```

Comment:
How come you don't need to recurse over this?

Reply:
Good catch, I should be 👍